include/rx/ranges.hpp

```diff
@@ -219,41 +219,48 @@ constexpr auto operator|(LHS&& lhs, RHS&& rhs) noexcept {
     return std::forward<RHS>(rhs)(std::forward<LHS>(lhs));
 }
 
-struct input_range_iterator_end {};
 template <class R>
 struct input_range_iterator {
     R range;
+    constexpr input_range_iterator() noexcept = default;
     template <class Arg>
-    constexpr explicit input_range_iterator(Arg&& range) noexcept
-        : range(std::forward<Arg>(range)) {}
-    constexpr input_range_iterator(input_range_iterator&&) noexcept = default;
-    constexpr input_range_iterator(const input_range_iterator&) noexcept = default;
-    constexpr input_range_iterator& operator=(input_range_iterator&&) noexcept = default;
-    constexpr input_range_iterator& operator=(const input_range_iterator&) noexcept = default;
+    constexpr explicit input_range_iterator(Arg&& range)
+        noexcept(std::is_nothrow_constructible_v<R, std::in_place_t, std::add_rvalue_reference_t<Arg>>)
+        : range(std::in_place, std::forward<Arg>(range)) {}
+    constexpr input_range_iterator(input_range_iterator&&) noexcept(std::is_nothrow_move_constructible_v<R>) = default;
+    constexpr input_range_iterator(const input_range_iterator&) noexcept(std::is_nothrow_copy_constructible_v<R>) = default;
+    constexpr input_range_iterator& operator=(input_range_iterator&&) noexcept(std::is_nothrow_move_assignable_v<R>) = default;
+    constexpr input_range_iterator& operator=(const input_range_iterator&) noexcept(std::is_nothrow_copy_assignable_v<R>) = default;
 
-    constexpr bool operator==(input_range_iterator_end) const {
-        return range.at_end();
+    constexpr bool operator==(const input_range_iterator &other) const noexcept {
+        RX_ASSERT(!bool(other.range));
+        return !bool(range) || range->at_end();
```

Comment:
This seems like the wrong thing to do. It would often lead to `begin(range)` and `end(range)` each having their own copy of the input, which could be a `vector_range` holding on to heap memory. Or am I missing something?

Reply:
Ah I see you have modified the definition of `end()` to return a default-constructed range object instead. That might be OK, but still more expensive than the `input_range_iterator_end` sentinel object. Is there an issue with having heterogeneous iterator types?

Reply:
>  Is there an issue with having heterogeneous iterator types?

Yes. In C++17 `for (V v : r) …`  is syntactic sugar for

```c++
auto &&iter = r;
auto _begin = std::begin(iter), _end = std::end(iter);
for (; _begin != _end; ++_begin) {
    V v = *_begin;
    …;
}
```

That is, _begin and _end have the same type. This definition was relaxed in C++20, where the second row was replaced with `auto _begin = std::begin(iter); auto _end = std::end(iter);`.

Reply:
No, that requirement was [already relaxed](https://en.cppreference.com/w/cpp/language/range-for) in C++17. :-)

The only thing that C++20 adds is an optional "init" statement.

Reply:
It is a minor thing in my opinion, since I think non-default-constructible comparison functions basically don't exist in the wild, but it may be best to not artificially introduce the requirement.

Reply:
I used `RX_OPTIONAL<R> range`, so R does not need to be default constructible. Should I use some other member name to make it clear that `range` is kinda "special"?